### ISSUE TITLE
How to detect if the player pick up an specific item (medkit, ammo, armor)

### ISSUE THREAD

**Issue by flyers** [mwydmuch/ViZDoom#229]
Hello,
I want to detect if the player pick up a medkit at every time step. If the agent is not in full health, then this can be derived by calculating the difference of GameVariable.HEALTH. However, I do not know how to do it when the health is already full. 
I have a rough idea that I need to change the ACS scripts to insert to a variable so that the variable will be updated when medkit is picked up. However, I am new to ACS and do not know where to start. Could anyone provide an example on this based on the cig.wad script?
Thanks.

**Comment by wjaskowski** [comment]
I am not an expert in ACS but I believe you need to workaround it. E.g., if
you detect full health in ACS, artificially decrease it to 99. You could
also change the MaxHealth (
https://zdoom.org/wiki/Actor_properties#Player.MaxHealth)


--
Wojciech Jaśkowski, Ph.D.
Laboratory of Intelligent Decision Support Systems
Institute of Computing Science
Poznan University of Technology, Poland
http://www.cs.put.poznan.pl/wjaskowski

**Comment by flyers** [comment]
@wjaskowski Thanks for your suggestion. I do not know how to detect full health in ACS. I tried to create a new player like follow:
```python
ACTOR MyPlayer : DoomPlayer
{
  Player.MaxHealth 1000
}
```
I put the above lines into the decorate part in cig.wad. But it did not have any effect. I think I must miss a few things.

**Comment by ebonyclock** [comment]
@flyers There is more elegant solution for this, create a custom healthpacks in DECORATE file which can be picked up when hp is full. I don't have time right now so I will elaborate in the evening.

**Comment by flyers** [comment]
@ebonyclock looking forward to your suggestion. Thanks.

**Comment by ebonyclock** [comment]
@flyers  sorry for such a long delay with reply.

Here is a simple solution:
- add a DECORATE file to the wad (it must be the first file) 
- add a definition of medikit object which is the same as the original https://zdoom.org/wiki/Classes:Medikit
- add "+INVENTORY.ALWAYSPICKUP"
- save, it seems that no compilation is needed

Here is the modified version:
ACTOR Medikit : Health
{
  Inventory.Amount 25
  Inventory.PickupMessage "$GOTMEDIKIT" // "Picked up a medikit."
  Health.LowMessage 25, "$GOTMEDINEED" // "Picked up a medikit that you REALLY need!"
  +INVENTORY.ALWAYSPICKUP
  States
  {
  Spawn:
    MEDI A -1
    Stop
  }
}

if you wanted you could change the name of the thing to e.g. "AlwaysPickableMedikit" bet then you'd need to change thing classes on the map.

**Comment by flyers** [comment]
@ebonyclock Thanks very much for your suggestion. I have tried it and it worked. The player can now pickup the medkit when the health is full.

However, I still can't figure out how to detect whether a medkit is picked up in a python script, since the health won't grow further. Is there anything like GameVariable.MEDKIT which stores the number of collected medkits? Or do I have to manually increase the maximum health limit of player so that I can infer by health differences?
Thanks very much.

**Comment by ebonyclock** [comment]
@flyers Look at this file: https://github.com/ebonyclock/vizdoom_cig2017/blob/master/cig2017_shaping.wad 
When player picks up a health pack USER1 game variable is increased.

**Comment by flyers** [comment]
@ebonyclock Thanks! I understand now.
I tried this and I found that the script can't count the host agent whether collected an medkit, where items for the added bots can be correctly calculated. Why is that?

**Comment by ebonyclock** [comment]
@flyers 
I don't understand what you wrote. Could you rephrase it and be more specific?

**Comment by flyers** [comment]
```python
import argparse
import os
import cv2
from vizdoom import *

game = DoomGame()
game.load_config('cig_shaping.cfg') # cig_shaping is the same as cig.cfg except it uses cig2017_shaping.wad
game.set_doom_map('map01')  # Limited deathmatch.

game.add_game_args("-host 1 -deathmatch +sv_spawnfarthest 1 "
                   "+timelimit 10.0 +sv_forcerespawn 1 +sv_noautoaim 1 +sv_respawnprotect 1 +sv_nocrouch 1 "
                   "+viz_respawn_delay 2 +viz_nocheat 0 +sv_unlimited_pickup 1")

game.add_game_args("+name AI +colorset 0")
game.set_screen_format(ScreenFormat.RGB24)

# Sets resolution for all buffers.
game.set_screen_resolution(ScreenResolution.RES_640X480)

# Enables depth buffer.
game.set_depth_buffer_enabled(True)

# Enables labeling of in game objects labeling.
game.set_labels_buffer_enabled(True)

game.set_mode(Mode.ASYNC_SPECTATOR)

game.init()
game.send_game_command("removebots")
for i in range(7):
    game.send_game_command("addbot")

# Play until the game (episode) is over.
cnt = 0
while not game.is_episode_finished():

    s = game.get_state()
    cnt += 1
    game.advance_action()

    if game.is_player_dead():
        game.respawn_player()

    print("medkit player 1:", game.get_game_variable(GameVariable.USER1))
    print("medkit player 2:", game.get_game_variable(GameVariable.USER2))
    print("medkit player 3:", game.get_game_variable(GameVariable.USER3))
    print("medkit player 4:", game.get_game_variable(GameVariable.USER4))
    print("medkit player 5:", game.get_game_variable(GameVariable.USER5))
    print("medkit player 6:", game.get_game_variable(GameVariable.USER6))
    print("medkit player 7:", game.get_game_variable(GameVariable.USER7))
    print("medkit player 8:", game.get_game_variable(GameVariable.USER8))
    print("medkit player 9:", game.get_game_variable(GameVariable.USER9))
    print("medkit player 10:", game.get_game_variable(GameVariable.USER10))
    print("medkit player 11:", game.get_game_variable(GameVariable.USER11))
    print("medkit player 12:", game.get_game_variable(GameVariable.USER12))
    print("medkit player 13:", game.get_game_variable(GameVariable.USER13))
    print("medkit player 14:", game.get_game_variable(GameVariable.USER14))
    print("medkit player 15:", game.get_game_variable(GameVariable.USER15))
    print("medkit player 16:", game.get_game_variable(GameVariable.USER16))
    # print("Frags:", game.get_game_variable(GameVariable.FRAGCOUNT))
    # print("Health:", game.get_game_variable(GameVariable.HEALTH))
    # print("Ammo:", game.get_game_variable(GameVariable.AMMO5))
    # print("Performed Action:", game.get_last_action())
    # print("X:", game.get_game_variable(GameVariable.POSITION_X))
    # print("Y:", game.get_game_variable(GameVariable.POSITION_Y))

game.close()

```
The above python code uses cig2017_shaping.wad to setup an deathmatch where the human player plays against other 7 built-in bots. I found that the USER? variable can only capture the medkits of the 7 added built-in AI bots, but not the human player. Even if I change to PlayerNumber() condition argument in the ACS script 4, the collected medkits of the human player are not counted.
@ebonyclock

**Comment by flyers** [comment]
I recently found that I made a stupid mistake. The modified ACS must be compiled first to make effect. This issue can be closed now. Thanks.

**Comment by sjtuytc** [comment]
So how to compile the modified ACS file?

**Comment by Miffyli** [comment]
@sjtuytc 

In the SLADE editor, right click on the ACS script entry and select "Scripts -> Compile ACS". You need to set path to ACC binary for this to work (easy to compile [from sources](https://github.com/rheit/acc) on Linux, or use [precompiled ones for Windows](https://devbuilds.drdteam.org/ACC/))
